public/app/core/services/backend_srv.ts

```diff
@@ -1,57 +1,119 @@
+import { AppEvents } from '@grafana/data';
+
 import appEvents from 'app/core/app_events';
 import config from 'app/core/config';
 import { DashboardModel } from 'app/features/dashboard/state/DashboardModel';
 import { DashboardSearchHit } from 'app/types/search';
-import { ContextSrv } from './context_srv';
-import { FolderInfo, DashboardDTO, CoreEvents } from 'app/types';
-import { BackendSrv as BackendService, getBackendSrv as getBackendService, BackendSrvRequest } from '@grafana/runtime';
-import { AppEvents } from '@grafana/data';
+import { CoreEvents, DashboardDTO, FolderInfo } from 'app/types';
+import { ContextSrv, contextSrv } from './context_srv';
+import { coreModule } from 'app/core/core_module';
+import { Emitter } from '../utils/emitter';
 
 export interface DatasourceRequestOptions {
   retry?: number;
   method?: string;
   requestId?: string;
-  timeout?: angular.IPromise<any>;
+  timeout?: Promise<any>;
   url?: string;
-  headers?: { [key: string]: any };
+  headers?: Record<string, any>;
   silent?: boolean;
-  data?: { [key: string]: any };
+  data?: Record<string, any>;
+}
+
+interface FetchResponseProps {
+  message?: string;
+}
+
+interface ErrorResponseProps extends FetchResponseProps {
+  status?: string;
+  error?: string | any;
+}
+
+export interface FetchResponse<T extends FetchResponseProps = any> {
+  status: number;
+  statusText: string;
+  ok: boolean;
+  data: T;
+}
+
+interface SuccessResponse extends FetchResponseProps, Record<any, any> {}
+
+interface DataSourceSuccessResponse<T extends {} = any> {
+  data: T;
+}
+
+interface ErrorResponse<T extends ErrorResponseProps = any> {
+  status: number;
+  statusText?: string;
+  isHandled?: boolean;
+  data: T | string;
+  cancelled?: boolean;
+}
+
+function serializeParams(data: Record<string, any>): string {
+  return Object.keys(data)
+    .map(key => {
+      const value = data[key];
+      if (Array.isArray(value)) {
+        return value.map(arrayValue => `${encodeURIComponent(key)}=${encodeURIComponent(arrayValue)}`).join('&');
+      }
+      return `${encodeURIComponent(key)}=${encodeURIComponent(value)}`;
+    })
+    .join('&');
+}
+
+export interface BackendSrvDependencies {
+  fromFetch: (input: string | Request, init?: RequestInit) => Observable<Response>;
+  appEvents: Emitter;
+  contextSrv: ContextSrv;
+  logout: () => void;
 }
 
 export class BackendSrv implements BackendService {
-  private inFlightRequests: { [key: string]: Array<angular.IDeferred<any>> } = {};
+  private inFlightRequests: Subject<string> = new Subject<string>();
   private HTTP_REQUEST_CANCELED = -1;
   private noBackendCache: boolean;
+  private dependencies: BackendSrvDependencies = {
+    fromFetch: fromFetch,
+    appEvents: appEvents,
+    contextSrv: contextSrv,
+    logout: () => {
+      window.location.href = config.appSubUrl + '/logout';
+    },
+  };
+
+  constructor(deps?: BackendSrvDependencies) {
```

Comment:
Can't see this constructor being used anywhere? Maybe we can get rid of it?

Reply:
Used in tests, I hope?

Reply:
Ah, I see now. Diff was hidden by default for backend_srv test henc cmd+f didn't work